plugins/vbe/fu-vbe-simple-device.c

```diff
@@ -0,0 +1,844 @@
+			    self->devname,
+			    self->fd,
+			    (uintmax_t)self->area_start,
+			    strerror(errno));
+		return NULL;
+	}
+
+	chunks = g_ptr_array_new_with_free_func((GDestroyNotify)g_bytes_unref);
+
+	for (offset = 0; offset < self->area_size; offset += blksize) {
+		g_autoptr(GBytes) chunk = NULL;
+		gsize toread;
+
+		buf = g_malloc(blksize);
+		toread = blksize;
+		if ((off_t)toread + offset > self->area_size)
+			toread = self->area_size - offset;
+		g_debug("read %zx", toread);
+		ret = read(self->fd, buf, toread);
+		if (ret < 0) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_READ,
+				    "Cannot read file '%s' (%s)",
+				    self->devname,
+				    strerror(errno));
+			g_free(buf);
+			return NULL;
+		}
+		chunk = g_bytes_new_take(buf, ret);
+		g_ptr_array_add(chunks, g_steal_pointer(&chunk));
+	}
+
+	out = fu_dfu_utils_bytes_join_array(chunks);
+	g_debug("Total bytes read from device: %#zx\n", g_bytes_get_size(out));
+
+	return out;
+}
+
+static void
+fu_vbe_simple_device_set_progress(FuDevice *self, FuProgress *progress)
+{
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 0, "detach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 100, "write");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 0, "attach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 0, "reload");
+}
+
+static void
+fu_vbe_simple_device_init(FuVbeSimpleDevice *self)
+{
+	fu_device_set_name(FU_DEVICE(self), "simple");
+	fu_device_add_guid(FU_DEVICE(self), "bb3b05a8-ebef-11ec-be98-d3a15278be95");
+	fu_device_set_vendor(FU_DEVICE(self), "U-Boot");
+	fu_device_add_vendor_id(FU_DEVICE(self), "VBE:U-Boot");
+	fu_device_set_version_format(FU_DEVICE(self), FWUPD_VERSION_FORMAT_TRIPLET);
+	fu_device_set_version_lowest(FU_DEVICE(self), "0.0.1");
+}
+
+FuDevice *
+fu_vbe_simple_device_new(FuContext *ctx,
+			 const gchar *vbe_method,
+			 const gchar *fdt,
+			 gint node,
+			 const gchar *vbe_dir)
+{
+	return FU_DEVICE(g_object_new(FU_TYPE_VBE_SIMPLE_DEVICE,
+				      "context",
+				      ctx,
+				      "vbe-method",
+				      vbe_method,
+				      "fdt",
+				      fdt,
+				      "node",
+				      node,
+				      "vbe-dir",
+				      vbe_dir,
+				      NULL));
+}
+
+static void
+fu_vbe_simple_device_constructed(GObject *obj)
+{
+	FuVbeSimpleDevice *self = FU_VBE_SIMPLE_DEVICE(obj);
+	self->vbe_fname =
+	    g_build_filename(fu_vbe_device_get_dir(FU_VBE_DEVICE(self)), "simple.dtb", NULL);
+}
+
+static void
+fu_vbe_simple_device_get_property(GObject *obj, guint prop_id, GValue *value, GParamSpec *pspec)
+{
+	FuVbeSimpleDevice *self = FU_VBE_SIMPLE_DEVICE(obj);
+	switch (prop_id) {
+	case PROP_DEVNAME:
+		g_value_set_string(value, self->devname);
+		break;
+	case PROP_AREA_START:
+		g_value_set_int64(value, self->area_start);
+		break;
+	case PROP_AREA_SIZE:
+		g_value_set_int64(value, self->area_size);
+		break;
```

Comment:
@sjg20 I'm a bit confused why these are GObject properties; you're not planning to subclass FuVbeSimpleDevice are you? If so, the device needs to be derivable, not final.

Reply:
No I am not planning to subclass it. I added these so that I can call the constructor with their values. Also I hope that these values might be visible to the fwupd client?

Reply:
OK, so what should we do here?

Reply:
Do you *need* the area-start values in the client? If so, why?

Reply:
It allows people to see the status of things, in this case where the firmware is located. With the A/B/recovery version, when I do it, we'll want to show the current state of things in more detail, e.g. what versions are in slots A and B. I suppose that could be done by reading the settings .dtb file directly, but it seems better to use fwupd if possible.

Reply:
The way we've done this so far is to expose each of the A/B partitions as separate devices, with the "other" partition being marked as non-updatable and as a child to the "active" partition. In the Redfish server case it seems to work really well.

Reply:
OK that sounds good. Once there is a successful boot we need to copy the firmware from B to A and roll forward the TPM.